src/units/abilities.cpp

```diff
@@ -1591,6 +1587,17 @@ bool attack_type::special_active_impl(const_attack_ptr self_attack, const_attack
 	const_attack_ptr att_weapon = is_attacker ? self_attack : other_attack;
 	const_attack_ptr def_weapon = is_attacker ? other_attack : self_attack;
 
+	// Filter firststrike here, if both units have first strike then the effects cancel out. Only check
+	// the opponent if "whom" is the defender, otherwise this leads to infinite
+	// recursion.
+	if (tag_name == "firststrike") {
+		// True if "whom" corresponds to "self", false if "whom" is "other"
+		bool whom_is_self = (whom == AFFECT_SELF) || ((whom == AFFECT_EITHER) && special_affects_self(special, is_attacker));
+		bool whom_is_defender = whom_is_self ? !is_attacker : is_attacker;
+		if (whom_is_defender && att_weapon && att_weapon->has_special_or_ability("firststrike"))
+			return false;
+	}
```

Comment:
Not that it should be done in this PR, but I wonder if these should be converted to a (string) enum rather than putting the weapon special tag name string directly in a bunch of places.